src/weights.jl

```diff
@@ -193,9 +195,87 @@ pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
     end
 end
 
+# NOTE: A custom constructor is needed because exponential weights must
+# sum to 1, so we want to enforce that on construction of the type
+@weights ExponentialWeights false
+
+"""
+    ExponentialWeights(vs)
+
+Construct an `ExponentialWeights` vector with weight values `vs`, which must sum to 1.
+
+Exponential weights are a common form of temporal weights which assign exponentially
+greater weight to past observations, which in this case corresponds to the tail end of
+the vector.
+"""
+function ExponentialWeights(vs::V) where {T<:Real, V<:AbstractVector{T}}
+    s = sum(vs)
+    s ≈ one(T) || throw(ArgumentError("weight values do not sum to 1 (got $s)"))
+    ExponentialWeights{T, T, V}(vs, s)
+end
+
+"""
+    eweights(n, λ)
+
+Construct an [`ExponentialWeights`](@ref) vector with length `n`,
+where each element in position ``i`` is set to ``λ (1 - λ)^{1 - i}``.
+The entire set of weights are then normalized to sum to 1.
+
+``λ`` is a smoothing factor or rate parameter such that ``0 < λ \\leq 1``.
+As this value approaches 0, the resulting weights will be almost equal,
+while values closer to 1 will put greater weight on the tail elements of the vector.
+
+# Examples
+
+```julia-repl
+julia> eweights(10, 0.3)
+10-element ExponentialWeights{Float64,Float64,Array{Float64,1}}:
+ 0.012458
+ 0.0177971
+ 0.0254245
+ 0.0363207
+ 0.0518867
+ 0.0741238
+ 0.105891
+ 0.151273
+ 0.216104
+ 0.308721
+```
+"""
+function eweights(n::Integer, λ::Real)
+    n > 0 || throw(ArgumentError("cannot construct exponential weights of length < 1"))
+    0 < λ <= 1 || throw(ArgumentError("smoothing factor must be between 0 and 1"))
+    w0 = map(i -> λ * (1 - λ)^(1 - i), 1:n)
```

Comment:
`setindex!` is defined for all `AbstractWeights`, but if we don't store the values then we can't `setindex!`. I still think weights should be immutable (though I know that discussion has occurred already), but as long as the other weight types are, `ExponentialWeights` should be as well, IMO.

Reply:
Well, the difference is that other types can be made mutable for free, while for exponential weights this requires allocating a vector. So I think the inconsistency is worth it.

Reply:
`ExponentialWeights` was literally Rory's use case for mutable weights though :laughing:

Reply:
How so?

Reply:
See e.g. https://github.com/JuliaStats/StatsBase.jl/issues/397#issuecomment-405793482

Reply:
If reversing exponential weights is a common need, better add an argument and field to do that?

Reply:
@rofinn, see above--thoughts?

Reply:
FWIW, I don't think we should compute the weights on the fly as this seems more appropriate for an iterable rather than a subtype of `AbstractArray`.

Reply:
Why? Ranges compute values on the fly on they are `AbstractArray`. As long as it's efficient, I don't see the problem.

Reply:
Yes, but I'm not likely to want to multiple a data matrix by a range.

Reply:
I don't get it. What's the problem?

Reply:
My point is just that you're going to need to allocate the vector to do anything useful with the weights anyways, so why bother special casing them just because you could compute the weights on the fly.

Reply:
Well at least you'll avoid storing a full vector all the time, and you save one copy. Also it's not really special casing: exponential weights are already special since they can be computed programmatically, contrary to other weights which are arbitrary. They could be complemented with unit weights, weights that sum to one or other kinds of weights in the future.

Reply:
A new reason to store all the vector might be if we want exponential weights computed based on timestamps. For example)

```
eweights(time::Vector{T}, range::AbstractRange{T}, λ) where T<:AbstractDateTime
```

In this case, the `time` vector represents a set of observation timestamps relative to a datetime `range`. If we have missing observations `time`s in that range then we want to return weights which reflect that.

Reply:
On a separate note, is the `AbstractDateTime` method is something we might want to include here?

Reply:
Maybe. How would you define that method?

However, exponential weights more and more appear as a different kind than other weights. The main goal of supporting the current three weights types is to apply corrections automatically (e.g. for `var`). That's not the case for exponential weights since no correction applies (https://github.com/JuliaStats/StatsBase.jl/pull/401#discussion_r214542438). So `eweights` just seems like a convenience function to generate a `Vector{Float64}`, and I'm not sure there's any reason to define a new `AbstractWeight` type, especially since https://github.com/JuliaLang/julia/pull/31395 makes it unnecessary for dispatch by using a keyword argument.

Reply:
> However, exponential weights more and more appear as a different kind than other weights.

Agreed. I often think that having a concept such as "importance" weights would be useful. I think STATA already has this?

> I'm not sure there's any reason to define a new AbstractWeight type, especially since JuliaLang/julia#31395 makes it unnecessary for dispatch by using a keyword argument.

I think it's still helpful to store the precomputed sum and it's useful for distinguishing weights from other array types even if base switches to using a keyword weights argument.

Reply:
As I noted in a previous thread, importance weights in Stata are just arbitrary weights, i.e. equivalent to our `Weights`.

Reply:
Okay, sorry, it's been a while. Would you be okay if we changed this PR to have `eweights` return `Weights`?

Reply:
Yes, in that case it would make more sense to allocate a `Vector{Float64}`, since you would be allowed to modify values without creating a contradiction with the name of the type.

Reply:
Alright, I think the changes I've made provide a more general/flexible `eweights` method, without the bias correction assumptions.